src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState_SourceGenerators.cs

```diff
@@ -26,52 +30,92 @@ internal sealed record SourceGeneratorMap(
         return map is null ? [] : map.SourceGenerators;
     }
 
+    /// <summary>
+    /// This method should only be called in a .net core host like our out of process server.
+    /// </summary>
     private static AnalyzerReference GetAnalyzerReference(ProjectState projectState, ISourceGenerator sourceGenerator)
     {
-        var map = GetSourceGeneratorMap(projectState.Language, projectState.AnalyzerReferences);
+        var map = GetSourceGeneratorMap(projectState);
         Contract.ThrowIfNull(map);
         return map.SourceGeneratorToAnalyzerReference[sourceGenerator];
     }
 
-    private static SourceGeneratorMap? GetSourceGeneratorMap(string language, IReadOnlyList<AnalyzerReference> analyzerReferences)
+    private static SourceGeneratorMap? GetSourceGeneratorMap(ProjectState projectState)
     {
-        var tupleOpt = s_languageToAnalyzerReferencesToSourceGeneratorsMap.FirstOrNull(static (t, language) => t.language == language, language);
-        if (tupleOpt is null)
+        if (!projectState.SupportsCompilation)
             return null;
 
-        var tuple = tupleOpt.Value;
-        return tuple.referencesToGenerators.GetValue(analyzerReferences, tuple.callback);
+        return s_projectStateToSourceGeneratorsMap.GetValue(projectState, ComputeSourceGenerators);
+
+        static SourceGeneratorMap ComputeSourceGenerators(ProjectState projectState)
+        {
+            using var generators = TemporaryArray<ISourceGenerator>.Empty;
+            using var _ = PooledDictionary<ISourceGenerator, AnalyzerReference>.GetInstance(out var generatorToAnalyzerReference);
+
+            foreach (var reference in projectState.AnalyzerReferences)
+            {
+                foreach (var generator in reference.GetGenerators(projectState.Language).Distinct())
+                {
+                    generators.Add(generator);
+                    generatorToAnalyzerReference.Add(generator, reference);
+                }
+            }
+
+            return new(generators.ToImmutableAndClear(), generatorToAnalyzerReference.ToFrozenDictionary());
+        }
+    }
+
+    public async Task<bool> HasSourceGeneratorsAsync(ProjectId projectId, CancellationToken cancellationToken)
+    {
+        var projectState = this.SolutionState.GetRequiredProjectState(projectId);
+
+        if (!s_hasSourceGeneratorsMap.TryGetValue(projectState, out var lazy))
+        {
+            // Extracted into local function to prevent allocations in the case where we find a value in the cache.
+            lazy = GetLazy(projectState);
+        }
+
+        return await lazy.GetValueAsync(cancellationToken).ConfigureAwait(false);
+
+        AsyncLazy<bool> GetLazy(ProjectState projectState)
+            => s_hasSourceGeneratorsMap.GetValue(
+                projectState,
+                projectState => AsyncLazy.Create(cancellationToken => ComputeHasSourceGeneratorsAsync(projectState, cancellationToken)));
```

Comment:
>projectState [](http://example.com/codeflow?start=16&length=12)

make static?

Reply:
this one is not actually staticable.  it needs to access the SolutionCompilationState so we can make the call to oop to figure things out.

Reply:
made partly static.

Reply:
I worded this poorly. I was asking about using the Arg overload of AsyncLazy.Create to prevent that closure allocation